alphagov/pay-connector
# PP-712 Investigate connector db connection pool and entity manager problems

## WHAT

_A brief description of the pull request:_

_Incident report - https://docs.google.com/document/d/1i0LYBAkaQ98Yzti9f6O0T61mCurIh4vLkl-uPD_wC3A/edit_
- **Problem 1:**
  In the integration we were getting inconsistent results for charge status and charge events. This was happening while the database healthcheck was running.
  The database healthcheck has been disabled which seem to have resolved the problem. 
  We need to do investigate what the root cause is.
- **Problem 2:**
  The test environment became non-responsive and when a thread dump was enforced it was noticed that we were exhausting our connection pool. 
  We need to investigate why/what exhausted the connection pool and come up with more healthchecks to enable monitoring similar issues.
## HOW

_Steps to test or reproduce:_
- There is a problem on how Guice Persist manages units of work.
- Requesting an entity manager outside of a unit of work, like we
  were doing with the HealthCheck, Guice Persist will implicitly
  start the unit of work for you, but because we are not telling it
  to terminate the unit of work or the method is not annotated with
  `@Transactional`, which will enforce a termination of the unit of
  work once existing the method, the entityManager that is responsible
  for dealing with that request at that moment falls into a "limbo".
  state and can never flush any cached entities so they can be reloaded again.
- As a side note, the JPAPersistenceService manages a collection of
  EntityManagers (making them available to both  Jersey and the Dropwizard
  Healthcheck Registry) so depending on the entity Manager that was
  responsible to handle the query we would see different results.
- We will provide some additional information about the current
  state of the connection pool. There are quite a few more things
  that can be added (distinguish between `idle` and `active`
  connections, what type of monitoring system will consume these
  stats, etc. This can be done a later stage once the work has been
  prioritised.
- Hit the `/healthcheck` endpoint to check the additional DB stats
## WHO

_Who should review this pull request:_
- [ ] WebOps
